### PR TITLE
Update boost deps for core and mss, track Sparta PR397

### LINKED ISSUE #35: Boost 1.81 not working (on macOS) - non-default boost location not supported by CMakeLists
Hi,

I've had some issues for Olympia to build on macOs (M1, running Ventura 13.1), not sure if the issue is present in other systems: default boost installed with ```brew install boost``` gives v1.81, which has issues in Phoenix which cause it to fail with duplicate symbols when linking (https://github.com/boostorg/phoenix/issues/111). (also Sparta build fails on this)

Just for other people running into this, here's my journey on this issue:

First of all: not to worry, I've also got boost v1.76 via homebrew, but it's not in a default location. So I had to build Sparta, Olympia and so also sta_lib with a non default boost location.

Then it got a bit complicated:
- Sparta build needs ```-DBoost_DIR=/opt/homebrew/Cellar/boost@1.76/1.76.0_3/lib/cmake/Boost-1.76.0``` to point it to the cake folder for the right boost version. (it does not, for some reason, use the cmake default included files to find boost)
- Olympia includes ```sparta-config.cmake``` so needs the same command line option for make to find Boost.
- But Olympia includes sty_lib in the build tree as a submodule, which does use the cmake default scripts, and they need another define on the command line: ```-DBOOST_ROOT=/opt/homebrew/Cellar/boost@1.76/1.76.0_3```

Which makes for a cmake command that gets a bit verbose:
```
cmake -S . -B Release -DCMAKE_BUILD_TYPE=Release -DSPARTA_BASE=$HOME/src/map/sparta \
   -DBOOST_ROOT=/opt/homebrew/Cellar/boost@1.76/1.76.0_3 \
   -DBoost_DIR=/opt/homebrew/Cellar/boost@1.76/1.76.0_3/lib/cmake/Boost-1.76.0
```

Which then passes cmake but fails at build, because boost is not added as an include for ```core``` and ```mss```.

Which I've fixed in the respective CMakeLists.txt, PR incoming soon.

But more generally: it would be nice to harmonise the boost config across Sparta, Olympia and stf_lib, so only 1 command line arg is needed to use it if it's in a non-default place. This avoids the double command line args needed in cmake.

I am willing to file a PR against Sparta as well to harmonise it to the standard cmake way of finding boost, but the comments in the CMakeLists.txt there seem to hint the custom way is there for (historic) reasons.

### PR DESCRIPTION
**Pull request by peter-d**
This fixes #35 and tracks the changes in Sparta [PR397](https://github.com/sparcians/map/pull/397)

### CODE REVIEW

**Comment by knute-sifive** [comment]
Ok, step 1 complete: MAP PR merged: https://github.com/sparcians/map/pull/397

**Comment by peter-d** [comment]
@knute-sifive I'm kind of stuck at this point: I have the HDF5 issue fixed, but with the CI pinned to sparta v1.1.1 this won't be able to pass...

Any advice? I don't want to break master CI either...

**Comment by knute-sifive** [comment]
Sorry, been trying to find some time to get back to this!  I can help if you get stuck.

**Comment by knute-sifive** [comment]
Ah, just caught up.  I'll take a look

**Comment by knute-sifive** [comment]
Seeing if I can reproduce locally...

**Comment by knute-sifive** [comment]
The issue on the release build is related to LTO.  Looking into the magic to make it work...

**Comment by knute-sifive** [comment]
Hazaa!

**Comment by knute-sifive** [comment]
Hey Peter, update your clone and see if a release and/or debug build works for you.

File: core/CMakeLists.txt
Comment (peter-d) [inline core/CMakeLists.txt]: Since I don't do a conda install of sparta, this fails for me: sparta  include paths and link paths are not passed properly.

**Comment by peter-d** [comment]
I had to put back a few target_link_libraries to find sparta (I don't have it installed in conda env so it's not picked up automatically in the default paths).

Since that does not break CI, this looks good to go to me, but it could do with a review especially to check:
- is it ok to pull in mavis and stf submodules from current masters? I updated the references
- the environment.yml file was coming from a `conda list` but I don't run in conda, so I manually edited the file. It currently works, but maybe is not what we want.

**Comment by knute-sifive** [comment]
> I had to put back a few target_link_libraries to find sparta

That's so weird that's required for the core library build.  Curious -- what didn't link?  The issue I have with adding sparta as a link library is that I believe this causes the core archive to blow up in size.  I think.

> is it ok to pull in mavis and stf submodules from current masters?

I don't see problem with that.  Probably should use an stf tag and tag Mavis so we have consistency and not some random SHA.  I'll look into that on the next revision.

Updating the yaml file for Conda is fine -- the CI uses it, so if it passes, then it should good to go!

**Comment by knute-sifive** [comment]
> I had to put back a few target_link_libraries to find sparta

That's so weird that's required for the core library build.  Curious -- what didn't link?  The issue I have with adding sparta as a link library is that I believe this causes the core archive to blow up in size.  I think.

> is it ok to pull in mavis and stf submodules from current masters?

I don't see problem with that.  Probably should use an stf tag and tag Mavis so we have consistency and not some random SHA.  I'll look into that on the next revision.

Updating the yaml file for Conda is fine -- the CI uses it, so if it passes, then it should good to go!

**Comment by knute-sifive** [comment]
> I had to put back a few target_link_libraries to find sparta

That's so weird that's required for the core library build.  Curious -- what didn't link?  The issue I have with adding sparta as a link library is that I believe this causes the core archive to blow up in size.  I think.

> is it ok to pull in mavis and stf submodules from current masters?

I don't see problem with that.  Probably should use an stf tag and tag Mavis so we have consistency and not some random SHA.  I'll look into that on the next revision.

Updating the yaml file for Conda is fine -- the CI uses it, so if it passes, then it should good to go!

**Comment by peter-d** [comment]
I added `target_link_libraries(... SPARTA::sparta)` since that then automatically also sets the right include paths for sparta when compiling. But I did not really consider the impact it might have at linking stage...

Since it's all statically linked at the end, I guess it could be sufficient to just add the includes here?

File: CMakeLists.txt
Comment (knute-sifive) [inline CMakeLists.txt:31]: I'm wondering if uncommenting this line with fix your include issues.  Can you do that AND remove the target_link lines to see if you still have problems?

Comment (peter-d) [reply CMakeLists.txt:31]: It's what this line and the next in `FindSparta.cmake` are providing automatically when you add sparta as a link library:

https://github.com/sparcians/map/blob/6f0636f7a717238cd039be33f7975048449bb82c/sparta/cmake/FindSparta.cmake#L69

Comment (knute-sifive) [reply CMakeLists.txt:31]: Ok, I can reproduce the missing include problem, but the not boost one.  I'll admit that I don't fully grok how `cmake` brings in the goods from a package.  You'd think the `find_package` method would be enough!  But sparta must be installed in a "standard" location or something so that includes AND library paths are set up?  I don't get it...

Comment (peter-d) [reply CMakeLists.txt:31]: All `find_package` does is import a target for the library (as if it were defined elsewhere in the makefiles with `add_library`) but it won’t set up any default compile or link flags project wide, it will only add the relevant options to the targets when you explicitly say they depend on the imported target. In this case that means unless you explicitly add SPARTA::sparta it won’t pass the flags. Which you won’t notice if everything is installed in system wide include and lib paths, but it will break as soon as one of the components (either Sparta or the lobs it depends on) is not in that default path. 

That said, there could be a way to only add the include paths and maybe not the link flags, but I expect that to fail. 

Reason I see boost problems is because I have multiple boost versions via homebrew, and Sparta needs one that is not the default one. (There’s macOS issues with boost 1.81 command line arguments)

Comment (arupc) [reply CMakeLists.txt:31]: Thank you Peter and Knute for working through this. 

Trying to understand the latest status of this work: From the comments above, I understand at least the regression is not working yet. But is it just the regression or more than that? Would I be able to build riscv-perf-model / olympia with latest head of master of Sparta (which installs Sparta in the conda environment)?

Comment (peter-d) [reply CMakeLists.txt:31]: As long as Sparta (and it’s dependencies like boost hdf5 etc) are available in the default paths (line and includes) it just works. It’s just in case some dependencies are not in a default location , that the overridden paths are not passed through from Sparta to Olympia at this point… (eg I have a home brew boost version in macOS that is not the default one which gives me troubles when I want to this locally branch currently)

**Comment by peter-d** [comment]
@knute-sifive if you could review? 
For me this is good to go (if debug CI won't fail for a weird reason)
Thanks!

**Comment by knute-sifive** [comment]
Absolutely!  Yes, let's get this in by the weekend.

File: .github/workflows/ubuntu-build.yml
Comment (peter-d) [inline .github/workflows/ubuntu-build.yml:39]: Is this ready to merge with the PR in sparta now merged as well? This checks out map_v2.0.0 tag of sparta, but I guess that can stay?

Comment (knute-sifive) [reply .github/workflows/ubuntu-build.yml:39]: I'd like to make a release of Sparta with that change OR we can just use `map_v2` development branch (will cherry pick your merge).

Comment (klingaard) [reply .github/workflows/ubuntu-build.yml]: Pushed a change to map_v2.  Once regression passes, I'll merge this in!

**Review by knute-sifive — APPROVED** [review]

**Pull request merged**
